### PR TITLE
8273369: Computing micros between two instants unexpectedly overflows for some cases

### PR DESCRIPTION
**Pull request by naotoj**
Please review the fix to the issue. Avoiding overflow by not calling nanosUntil() directly, which will overflow beyond Long.MAX_VALUE difference in nano unit.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8273369](https://bugs.openjdk.java.net/browse/JDK-8273369): Computing micros between two instants unexpectedly overflows for some cases

### Reviewers
 * [Lance Andersen](https://openjdk.java.net/census#lancea) (@LanceAndersen - **Reviewer**)
 * [Roger Riggs](https://openjdk.java.net/census#rriggs) (@RogerRiggs - **Reviewer**)
 * [Joe Wang](https://openjdk.java.net/census#joehw) (@JoeWang-Java - **Reviewer**) ⚠️ Review applies to ccf73bf7417b93aaebeab1d2f079330e2f7cbfb7

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/5396/head:pull/5396` \
`$ git checkout pull/5396`

Update a local copy of the PR: \
`$ git checkout pull/5396` \
`$ git pull https://git.openjdk.java.net/jdk pull/5396/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 5396`

View PR using the GUI difftool: \
`$ git pr show -t 5396`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/5396.diff">https://git.openjdk.java.net/jdk/pull/5396.diff</a>

</details>

### CODE REVIEW

**Review by LanceAndersen — APPROVED** [review]
Looks OK to me

File: src/java.base/share/classes/java/time/Instant.java [hunk 1]
```diff
@@ -1164,6 +1164,12 @@ private long nanosUntil(Instant end) {
         return Math.addExact(totalNanos, end.nanos - nanos);
     }
 
+    private long microsUntil(Instant end) {
+        long secsDiff = Math.subtractExact(end.seconds, seconds);
+        long totalMicros = Math.multiplyExact(secsDiff, NANOS_PER_SECOND / 1000);
+        return Math.addExact(totalMicros, (end.nanos - nanos) / 1000);
```
Comment (RogerRiggs) [inline src/java.base/share/classes/java/time/Instant.java:1170]: Can you define NANOS_PER_MICRO, the others conversions use predefined constants.

**Review by RogerRiggs — APPROVED** [review]

File: src/java.base/share/classes/java/time/Instant.java (hunk 1 above)
Comment (naotoj) [reply src/java.base/share/classes/java/time/Instant.java:1174]: Defined it as a private field in `Instant`.

**Review by RogerRiggs — APPROVED** [review]

**Review by RogerRiggs — APPROVED** [review]

File: test/jdk/java/time/test/java/time/TestInstant.java [hunk 2]
```diff
@@ -96,4 +98,11 @@ public void test_epochMillis(String name, long millis) {
         assertEquals(millis, m, name);
     }
 
+    @Test
+    public void test_microsUntil() {
```
Comment (JoeWang-Java) [inline test/jdk/java/time/test/java/time/TestInstant.java:102]: A comment about the test might be helpful.

**Review by JoeWang-Java — APPROVED** [review]

File: test/jdk/java/time/test/java/time/TestInstant.java (hunk 2 above)
Comment (naotoj) [reply test/jdk/java/time/test/java/time/TestInstant.java:107]: Indeed. Added some comments to the test.

**Review by JoeWang-Java — APPROVED** [review]

**Review by RogerRiggs — APPROVED** [review]

**Review by jodastephen — COMMENTED** [review]
This change looks fine, but I think you also need a `millisUntil` private method to fix the identical overflow problem with millis (which might as well be fixed now).

**Comment by naotoj** [comment]
> This change looks fine, but I think you also need a `millisUntil` private method to fix the identical overflow problem with millis (which might as well be fixed now).

`until()` for millis simply subtracts its `toEpochMilli()` from the end Instant's, so I am not sure that would cause the same false `ArithmeticException`. Can you please elaborate more?

**Comment by jodastephen** [comment]
`toEpochMilli()` overflows at large/small values. Thus any attempt to calculate the difference between two very large instants would fail.

**Comment by naotoj** [comment]
> `toEpochMilli()` overflows at large/small values. Thus any attempt to calculate the difference between two very large instants would fail.

Thanks. Fixed.

**Review by RogerRiggs — APPROVED** [review]

**Review by LanceAndersen — APPROVED** [review]

**Comment by Blizzara** [comment]
Flagging here as I didn't see it mentioned in these comments or in the issue: this PR changed the behavior of `until()` for `MILLIS`. The earlier version (arguably buggily) did not consider the micros/nanos when computing the millis number, which differs from how the other units behaved. This new version considers also the smaller parts. To see the difference, run e.g. the following:

```
import java.time.Instant;
import java.time.temporal.ChronoUnit;

public class Main {
    public static void main(String[] args) {
        System.out.println("SECONDS: " + ChronoUnit.SECONDS.between(Instant.ofEpochSecond(1, 1), Instant.ofEpochSecond(2, 0)));
        System.out.println("MILLIS: " + ChronoUnit.MILLIS.between(Instant.ofEpochSecond(1, 1), Instant.ofEpochSecond(2, 0)));
        System.out.println("MICROS: " + ChronoUnit.MICROS.between(Instant.ofEpochSecond(1, 1), Instant.ofEpochSecond(2, 0)));
    }
}
```

With Corretto 17, this prints:
```
SECONDS: 0
MILLIS: 1000
MICROS: 999999
```
while Zulu 21 gives:
```
SECONDS: 0
MILLIS: 999
MICROS: 999999
```
(different JDKs but the change comes from this PR)

**Pull request closed without merging**